File: src/lib/businessTime.ts
/**
 * Nexo opera en horario de Ecuador/Colombia (America/Guayaquil o Bogotá,
 * UTC-5, sin horario de verano). Los servidores (p. ej. Vercel) suelen
 * correr con reloj en UTC, así que cualquier cálculo de "qué día es hoy" en
 * tiempo real debe desplazar el instante actual a este huso ANTES de leer
 * el día calendario — de lo contrario "hoy" puede adelantarse hasta 5 horas
 * respecto al día real del usuario (p. ej. domingo 9pm local ya es lunes
 * 2am UTC).
 *
 * Esto NO aplica a campos de fecha pura (p. ej. Task.endDate, almacenados
 * como medianoche UTC por convención) — esos ya representan un día
 * calendario sin ambigüedad y no deben desplazarse.
 */
export const BUSINESS_TZ_OFFSET_HOURS = 5;

/** UTC-midnight Date for the business-timezone calendar day that `instant` falls on. */
export function businessCalendarDay(instant: Date): Date {
  const shifted = new Date(instant.getTime() - BUSINESS_TZ_OFFSET_HOURS * 3600000);
  return new Date(Date.UTC(shifted.getUTCFullYear(), shifted.getUTCMonth(), shifted.getUTCDate()));
}

/**
 * The real UTC instant range spanning a business-timezone calendar day —
 * for comparing against genuine timestamps (e.g. TaskActivity.createdAt),
 * as opposed to date-only fields like Task.endDate.
 */
export function businessDayRealRange(calDay: Date): { start: Date; end: Date } {
  const start = new Date(
    Date.UTC(calDay.getUTCFullYear(), calDay.getUTCMonth(), calDay.getUTCDate(), BUSINESS_TZ_OFFSET_HOURS)
  );
  return { start, end: new Date(start.getTime() + 86400000 - 1) };
}

/** Lunes a viernes, evaluado sobre un Date UTC-medianoche (ver businessCalendarDay). */
export function isBusinessDay(calDay: Date): boolean {
  const dow = calDay.getUTCDay();
  return dow !== 0 && dow !== 6;
}

/**
 * Los `count` días laborables (lun-vie) más recientes, estrictamente anteriores a
 * `today`, en orden descendente (el más reciente primero). `today` debe ser un
 * Date UTC-medianoche (ver businessCalendarDay) — usado para acotar la ventana de
 * registro retroactivo de horas a los últimos N días laborables.
 */
export function previousBusinessDays(today: Date, count: number): Date[] {
  const days: Date[] = [];
  const cursor = new Date(today);
  while (days.length < count) {
    cursor.setUTCDate(cursor.getUTCDate() - 1);
    if (isBusinessDay(cursor)) days.push(new Date(cursor));
  }
  return days;
}

/**
 * El sábado y domingo del fin de semana inmediatamente anterior a `today`,
 * pero solo cuando `today` es lunes o martes — a partir del miércoles esos
 * dos días ya no deben aparecer como registrables. `today` debe ser un Date
 * UTC-medianoche (ver businessCalendarDay). Devuelve más reciente primero
 * (domingo, luego sábado), igual que `previousBusinessDays`.
 */
export function weekendGraceDays(today: Date): Date[] {
  const dow = today.getUTCDay();
  if (dow !== 1 && dow !== 2) return [];
  const monday = new Date(today);
  monday.setUTCDate(monday.getUTCDate() - (dow - 1));
  const sunday = new Date(monday);
  sunday.setUTCDate(sunday.getUTCDate() - 1);
  const saturday = new Date(monday);
  saturday.setUTCDate(saturday.getUTCDate() - 2);
  return [sunday, saturday];
}

/**
 * Motor único de validación para registro retroactivo: los `count` días
 * laborables más recientes (ver `previousBusinessDays`) más, si aplica, el
 * fin de semana inmediato anterior (ver `weekendGraceDays`). Usado tanto por
 * el date-picker del cliente como por la validación del servidor, para que
 * ambos lados acepten exactamente el mismo conjunto de fechas.
 */
export function retroactiveValidDates(today: Date, businessDayCount: number): Date[] {
  return [...previousBusinessDays(today, businessDayCount), ...weekendGraceDays(today)].sort(
    (a, b) => b.getTime() - a.getTime()
  );
}

/**
 * "YYYY-MM-DD" (de un `<input type="date">`, sin componente horario) -> Date
 * UTC-medianoche. Antes de Sprint D estaba duplicado idéntico en el registro
 * retroactivo de Tareas y en las actividades de Proyectos — ver
 * docs/AUDIT_LOG.md § Sprint D.
 */
export function parseDateOnly(value: string): Date | null {
  if (!/^\d{4}-\d{2}-\d{2}$/.test(value)) return null;
  const [y, m, d] = value.split("-").map(Number);
  const parsed = new Date(Date.UTC(y, m - 1, d));
  return Number.isNaN(parsed.getTime()) ? null : parsed;
}

File: backend/apps/tasks/business_time.py
"""Helpers de tiempo de negocio — portados 1:1 de `src/lib/businessTime.ts`
y `src/lib/timeOverlap.ts` del Next.js legacy (Fase 3b, ver
docs/AUDIT_LOG.md § 2026-08-07). Puros: sin dependencias de Django."""

import re
from datetime import date, datetime, timedelta, timezone

BUSINESS_TZ_OFFSET_HOURS = 5

_TIME_RE = re.compile(r"^(\d{2}):(\d{2})$")


def business_calendar_day(instant: datetime) -> date:
    """El día calendario de negocio (huso desplazado `BUSINESS_TZ_OFFSET_HOURS`
    respecto a UTC) al que pertenece `instant`."""
    shifted = instant - timedelta(hours=BUSINESS_TZ_OFFSET_HOURS)
    return shifted.date()


def business_day_real_range(cal_day: date) -> tuple[datetime, datetime]:
    """El rango real de instantes UTC que abarca el día de negocio
    `cal_day` — para comparar contra timestamps reales
    (`TaskActivity.created_at`), no contra fechas sueltas."""
    start = datetime(
        cal_day.year, cal_day.month, cal_day.day, BUSINESS_TZ_OFFSET_HOURS, tzinfo=timezone.utc
    )
    end = start + timedelta(days=1) - timedelta(milliseconds=1)
    return start, end


def is_business_day(cal_day: date) -> bool:
    """Lunes a viernes."""
    return cal_day.weekday() < 5


def previous_business_days(today: date, count: int) -> list[date]:
    """Los `count` días laborables (lun-vie) más recientes, estrictamente
    anteriores a `today`, en orden descendente (el más reciente primero)
    — acota la ventana de registro retroactivo de horas."""
    days: list[date] = []
    cursor = today
    while len(days) < count:
        cursor = cursor - timedelta(days=1)
        if is_business_day(cursor):
            days.append(cursor)
    return days


def weekend_grace_days(today: date) -> list[date]:
    """El sábado y domingo del fin de semana inmediatamente anterior a
    `today`, solo cuando `today` es lunes (weekday=0) o martes
    (weekday=1) — a partir del miércoles esos días ya no son
    registrables. Devuelve más reciente primero (domingo, luego
    sábado)."""
    dow = today.weekday()
    if dow not in (0, 1):
        return []
    monday = today - timedelta(days=dow)
    sunday = monday - timedelta(days=1)
    saturday = monday - timedelta(days=2)
    return [sunday, saturday]


def retroactive_valid_dates(today: date, business_day_count: int) -> list[date]:
    """Motor único de validación para registro retroactivo: los
    `business_day_count` días laborables más recientes más, si aplica, el
    fin de semana inmediato anterior."""
    return sorted(
        [*previous_business_days(today, business_day_count), *weekend_grace_days(today)], reverse=True
    )


_DATE_ONLY_RE = re.compile(r"^(\d{4})-(\d{2})-(\d{2})$")


def parse_date_only(value: str) -> date | None:
    """`"YYYY-MM-DD"` (de un `<input type="date">`) -> `date`, o `None` si
    el formato no es válido — a diferencia de `task_import.parse_date`
    (que también acepta seriales de Excel/formatos ambiguos), el registro
    retroactivo legacy usa exclusivamente este formato estricto."""
    match = _DATE_ONLY_RE.match(value)
    if not match:
        return None
    try:
        return date(int(match.group(1)), int(match.group(2)), int(match.group(3)))
    except ValueError:
        return None


def time_to_minutes(value: str) -> int | None:
    """`"HH:MM"` -> minutos desde medianoche, o `None` si el formato no es
    válido."""
    match = _TIME_RE.match(value)
    if not match:
        return None
    hours, minutes = int(match.group(1)), int(match.group(2))
    if hours > 23 or minutes > 59:
        return None
    return hours * 60 + minutes


def ranges_overlap(a_start: str, a_end: str, b_start: str, b_end: str) -> bool:
    """`True` si `[a_start,a_end)` se solapa con `[b_start,b_end)` — cubre
    solapamiento parcial, contención total en cualquier dirección y
    coincidencia exacta."""
    s1, e1, s2, e2 = (
        time_to_minutes(a_start),
        time_to_minutes(a_end),
        time_to_minutes(b_start),
        time_to_minutes(b_end),
    )
    if None in (s1, e1, s2, e2):
        return False
    return s1 < e2 and s2 < e1
